File: examples/c/hello.c
/*
 * Minimal C sample against the SQLRite FFI surface.
 *
 * Walks through:
 *   - opening an in-memory connection
 *   - CREATE TABLE + INSERT via sqlrite_execute
 *   - SELECT via sqlrite_query + sqlrite_step + sqlrite_column_*
 *   - explicit memory freeing (sqlrite_free_string, sqlrite_finalize,
 *     sqlrite_close)
 *   - a BEGIN / ROLLBACK transaction block
 *
 * Build + run:
 *   cd sqlrite-ffi && cargo build --release
 *   cc -I../sqlrite-ffi/include \
 *      -L../target/release -lsqlrite_c \
 *      examples/c/hello.c -o hello
 *   # macOS: also pass -Wl,-rpath,../target/release
 *   ./hello
 *
 * (On Linux, add `-Wl,-rpath,$PWD/../target/release` so the runtime
 * can find libsqlrite_c.so without LD_LIBRARY_PATH juggling.)
 */

#include <stdio.h>
#include <stdlib.h>
#include "sqlrite.h"

/* Small helper that prints the library's last error and exits. */
static void die(const char *what) {
    const char *err = sqlrite_last_error();
    fprintf(stderr, "%s: %s\n", what, err ? err : "(no error message)");
    exit(1);
}

int main(void) {
    struct SqlriteConnection *conn = NULL;
    if (sqlrite_open_in_memory(&conn) != Ok) {
        die("sqlrite_open_in_memory");
    }

    if (sqlrite_execute(conn,
            "CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT, age INTEGER);") != Ok) {
        die("CREATE TABLE");
    }
    if (sqlrite_execute(conn,
            "INSERT INTO users (name, age) VALUES ('alice', 30);") != Ok) {
        die("INSERT alice");
    }
    if (sqlrite_execute(conn,
            "INSERT INTO users (name, age) VALUES ('bob', 25);") != Ok) {
        die("INSERT bob");
    }

    /* Show the table. */
    struct SqlriteStatement *stmt = NULL;
    if (sqlrite_query(conn, "SELECT id, name, age FROM users;", &stmt) != Ok) {
        die("sqlrite_query");
    }

    int col_count = 0;
    sqlrite_column_count(stmt, &col_count);
    printf("Columns: ");
    for (int i = 0; i < col_count; i++) {
        char *name = NULL;
        sqlrite_column_name(stmt, i, &name);
        printf("%s%s", name, i + 1 < col_count ? ", " : "\n");
        sqlrite_free_string(name);
    }

    while (1) {
        enum SqlriteStatus st = sqlrite_step(stmt);
        if (st == Done) break;
        if (st != Row) die("sqlrite_step");

        int64_t id = 0;
        char *name = NULL;
        int64_t age = 0;
        sqlrite_column_int64(stmt, 0, &id);
        sqlrite_column_text(stmt, 1, &name);
        sqlrite_column_int64(stmt, 2, &age);
        printf("  row: id=%lld name=%s age=%lld\n",
               (long long)id, name, (long long)age);
        sqlrite_free_string(name);
    }
    sqlrite_finalize(stmt);

    /* Transaction that gets rolled back: the row doesn't survive. */
    sqlrite_execute(conn, "BEGIN;");
    sqlrite_execute(conn, "INSERT INTO users (name, age) VALUES ('phantom', 99);");
    printf("\nIn transaction: %d\n", sqlrite_in_transaction(conn));
    sqlrite_execute(conn, "ROLLBACK;");
    printf("After rollback: %d\n", sqlrite_in_transaction(conn));

    /* Count rows post-rollback — should be 2. */
    sqlrite_query(conn, "SELECT id FROM users;", &stmt);
    int count = 0;
    while (sqlrite_step(stmt) == Row) count++;
    printf("Row count after rollback: %d (expected 2)\n", count);
    sqlrite_finalize(stmt);

    sqlrite_close(conn);
    return 0;
}

File: examples/go/hello.go
// Minimal walkthrough of the SQLRite Go driver.
//
// Prerequisite: build libsqlrite_c once from the repo root:
//
//	cargo build --release -p sqlrite-ffi
//
// Then from inside examples/go/:
//
//	go run hello.go
//
// Shape mirrors the standard library's `database/sql` API — if you've
// used any Go SQL driver (pq, mysql, sqlite3), you already know how
// to drive this.
package main

import (
	"database/sql"
	"fmt"
	"log"

	// Side-effect import registers "sqlrite" as a sql.Open driver name.
	_ "github.com/joaoh82/rust_sqlite/sdk/go"
)

func main() {
	// Pass `:memory:` for a transient in-memory DB (matching SQLite
	// convention); pass a file path like "foo.sqlrite" for a file-
	// backed DB that auto-saves on every write.
	db, err := sql.Open("sqlrite", ":memory:")
	if err != nil {
		log.Fatalf("open: %v", err)
	}
	defer db.Close()

	mustExec(db, "CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT, age INTEGER)")
	mustExec(db, "INSERT INTO users (name, age) VALUES ('alice', 30)")
	mustExec(db, "INSERT INTO users (name, age) VALUES ('bob', 25)")
	mustExec(db, "INSERT INTO users (name, age) VALUES ('carol', 40)")

	// database/sql's Query + Scan — typed into Go primitives.
	fmt.Println("All users:")
	rows, err := db.Query("SELECT id, name, age FROM users")
	if err != nil {
		log.Fatalf("query: %v", err)
	}
	for rows.Next() {
		var id, age int64
		var name string
		if err := rows.Scan(&id, &name, &age); err != nil {
			log.Fatalf("scan: %v", err)
		}
		fmt.Printf("  %d: %s (%d)\n", id, name, age)
	}
	rows.Close()

	// Transactions via database/sql: Begin, issue statements, commit
	// or rollback. The engine treats this the same as `BEGIN` /
	// `COMMIT` / `ROLLBACK` statements via Exec — auto-save is
	// suppressed until the transaction commits, and Rollback
	// restores the pre-BEGIN snapshot.
	fmt.Println()
	tx, err := db.Begin()
	if err != nil {
		log.Fatalf("begin: %v", err)
	}
	mustTxExec(tx, "INSERT INTO users (name, age) VALUES ('phantom', 99)")
	fmt.Printf("Mid-transaction row count: %d\n", countRows(tx, "SELECT id FROM users"))

	if err := tx.Rollback(); err != nil {
		log.Fatalf("rollback: %v", err)
	}
	fmt.Printf("Post-rollback row count:   %d\n", countRows(db, "SELECT id FROM users"))
}

// Small helpers — `database/sql`'s idiomatic error-handling gets
// noisy in example code, so we factor out two tiny wrappers.

type queryer interface {
	Query(query string, args ...any) (*sql.Rows, error)
}

func mustExec(db *sql.DB, q string) {
	if _, err := db.Exec(q); err != nil {
		log.Fatalf("exec %q: %v", q, err)
	}
}

func mustTxExec(tx *sql.Tx, q string) {
	if _, err := tx.Exec(q); err != nil {
		log.Fatalf("tx.Exec %q: %v", q, err)
	}
}

func countRows(q queryer, sqlText string) int {
	rows, err := q.Query(sqlText)
	if err != nil {
		log.Fatalf("query: %v", err)
	}
	defer rows.Close()
	n := 0
	for rows.Next() {
		n++
	}
	return n
}
